src/client/debugger/jupyter/debugControllers.ts

```diff
@@ -0,0 +1,173 @@
+import { traceVerbose } from '../../common/logger';
+import { IConfigurationService } from '../../common/types';
+import { noop } from '../../common/utils/misc';
+import { Commands } from '../../datascience/constants';
+import { IKernel } from '../../datascience/jupyter/kernels/types';
+import { sendTelemetryEvent } from '../../telemetry';
+import { DebuggingTelemetry } from '../constants';
+import { DebuggingDelegate, IKernelDebugAdapter } from '../types';
+
+export class DebugCellController implements DebuggingDelegate {
+    constructor(
+        private readonly debugAdapter: IKernelDebugAdapter,
+        public readonly debugCell: NotebookCell,
+        private readonly kernel: IKernel,
+        private readonly commandManager: ICommandManager
+    ) {
+        sendTelemetryEvent(DebuggingTelemetry.successfullyStartedRunAndDebugCell);
+    }
+
+    public async willSendEvent(_msg: DebugProtocolMessage): Promise<boolean> {
+        return false;
+    }
+
+    public async willSendRequest(request: DebugProtocol.Request): Promise<void> {
+        if (request.command === 'configurationDone') {
+            await cellDebugSetup(this.kernel, this.debugAdapter, this.debugCell);
+
+            void this.commandManager.executeCommand('notebook.cell.execute', {
+                ranges: [{ start: this.debugCell.index, end: this.debugCell.index + 1 }],
+                document: this.debugCell.document.uri
+            });
+        }
+    }
+}
+
+export class RunByLineController implements DebuggingDelegate {
+    private lastPausedThreadId: number | undefined;
+
+    constructor(
+        private readonly debugAdapter: IKernelDebugAdapter,
+        public readonly debugCell: NotebookCell,
+        private readonly commandManager: ICommandManager,
+        private readonly kernel: IKernel,
+        private readonly settings: IConfigurationService
+    ) {
+        sendTelemetryEvent(DebuggingTelemetry.successfullyStartedRunByLine);
+    }
+
+    public continue(): void {
+        if (typeof this.lastPausedThreadId !== 'number') {
+            traceVerbose(`No paused thread, can't do RBL`);
+            return;
+        }
+
+        void this.debugAdapter.stepIn(this.lastPausedThreadId);
+    }
+
+    public stop(): void {
+        this.debugAdapter.disconnect();
+    }
+
+    public async willSendEvent(msg: DebugProtocolMessage): Promise<boolean> {
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        const anyMsg = msg as any;
+
+        if (anyMsg.content.event === 'stopped') {
+            this.lastPausedThreadId = anyMsg.content.body.threadId;
+            if (await this.handleStoppedEvent(this.lastPausedThreadId!)) {
+                this.trace('intercepted', JSON.stringify(anyMsg.content));
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    public async willSendRequest(request: DebugProtocol.Request): Promise<void> {
```

Comment:
Not sure i like this `willSendrequest`.
`will` implies there's a question, and I'd expect the response to be `true/false` or similar.
But I'm not seeing a return value, hence this looks more like a command to do somethign as opposed to a method used to inquire somethign.

Reply:
Never mind, read the comments in interface.
Do we need an interface

Reply:
Should I call them `onWill...`?

Reply:
And the fact that one returns a boolean while the other doesn't may or may not be bad, I can't decide